File: spec/game/match_event_batch_spec.lua
local match_event_batch = require("game.presentation.match_event_batch")
local t = require("spec.support.runner")

---@param id string
---@param y number
---@return RollbackWrappedMatchEvent
local function tip(id, y)
    return {
        id = id,
        tick = 12,
        domain = "match/tip",
        ordinal = 1,
        payload = { kind = "tip", x = 24, y = y, player = "home_keeper" },
    }
end

---@return RollbackEventDiff
local function empty_diff()
    return { added = {}, revoked = {}, replaced = {} }
end

t.describe("rollback match-event presentation batch", function()
    t.it("does not present a tip added then revoked in one render update", function()
        local event = tip("tip/old", 220)
        local added = empty_diff()
        added.added[1] = event
        local revoked = empty_diff()
        revoked.revoked[1] = event

        t.eq(#match_event_batch.surviving({ added, revoked }), 0)
    end)

    t.it("presents only the corrected tip after added then replaced", function()
        local before = tip("tip/old", 220)
        local after = tip("tip/new", 320)
        local added = empty_diff()
        added.added[1] = before
        local replaced = empty_diff()
        replaced.replaced[1] = { before = before, after = after }

        local events = match_event_batch.surviving({ added, replaced })
        t.eq(#events, 1)
        t.eq(events[1].kind, "tip")
        t.eq(events[1].y, 320)
    end)
end)

File: ts/packages/presentation/src/match_event_batch.spec.ts
// Ported from spec/game/match_event_batch_spec.lua.

import { describe, expect, it } from "vitest";
import { matchEventBatch, type RollbackEventDiff, type RollbackWrappedEvent } from "./match_event_batch.ts";

interface TipEvent {
  readonly kind: "tip";
  readonly x: number;
  readonly y: number;
  readonly player: string;
}

function tip(id: string, y: number): RollbackWrappedEvent<TipEvent> {
  return {
    id,
    tick: 12,
    domain: "match/tip",
    ordinal: 1,
    payload: { kind: "tip", x: 24, y, player: "home_keeper" },
  };
}

function emptyDiff(): RollbackEventDiff {
  return { added: [], revoked: [], replaced: [] };
}

describe("rollback match-event presentation batch", () => {
  it("does not present a tip added then revoked in one render update", () => {
    const event = tip("tip/old", 220);
    const added: RollbackEventDiff = { ...emptyDiff(), added: [event] };
    const revoked: RollbackEventDiff = { ...emptyDiff(), revoked: [event] };

    expect(matchEventBatch.surviving([added, revoked])).toHaveLength(0);
  });

  it("presents only the corrected tip after added then replaced", () => {
    const before = tip("tip/old", 220);
    const after = tip("tip/new", 320);
    const added: RollbackEventDiff = { ...emptyDiff(), added: [before] };
    const replaced: RollbackEventDiff = { ...emptyDiff(), replaced: [{ before, after }] };

    const events = matchEventBatch.surviving([added, replaced]);
    expect(events).toHaveLength(1);
    const [survivor] = events;
    expect(survivor).toBeDefined();
    expect((survivor as unknown as TipEvent).kind).toBe("tip");
    expect((survivor as unknown as TipEvent).y).toBe(320);
  });
});
